specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,188 @@
+FlagEvaluationDetails<boolean> myBoolDetails = client.getBooleanDetails('bool-flag', false);
+
+// example detailed string flag evaluation
+FlagEvaluationDetails<string> myStringDetails = client.getStringDetails('string-flag', 'N/A', evaluationContext, options);
+
+// example detailed number flag evaluation
+FlagEvaluationDetails<number> myNumberDetails = client.getNumberDetails('number-flag', 75);
+
+// example detailed structure flag evaluation
+FlagEvaluationDetails<MyStruct> myStructDetails = client.getObjectDetails<MyStruct>('structured-flag', { text: 'N/A', percentage: 75 }, evaluationContext, options);
+
+```
+
+##### Requirement 1.10
+
+> The `evaluation details` structure's `value` field **MUST** contain the evaluated flag value.
+
+##### Condition 1.11
+
+> The language supports generics (or an equivalent feature).
+>
+> ##### Conditional Requirement 1.11.1
+>
+> > The `evaluation details` structure **SHOULD** accept a generic argument (or utilize an equivalent language feature) which indicates the type of the wrapped `value` field.
+
+##### Requirement 1.12
+
+> The `evaluation details` structure's `flag key` field **MUST** must contain the `flag key` argument passed to the detailed flag evaluation method.
+
+##### Requirement 1.13
+
+> The `evaluation details` structure's `variant` field **SHOULD** contain a identifier logically corresponding to the return flag value.
+
+For example, the flag value might be `3.14159265359`, and the variant field's value might be `"pi"`.
+
+##### Requirement 1.14
+
+> The `evaluation details` structure's `reason` field **SHOULD** indicate the semantic reason for the returned flag value. Possible values vary by provider, but may include `"TARGETING_MATCH"`, `"SPLIT"`, `"DISABLED"`, `"DEFAULT"`, `"UNKNOWN"` and `"ERROR"`.
+
+##### Requirement 1.15
+
+> In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` **SHOULD** be set to `"ERROR"`.
+
+##### Requirement 1.16
+
+> In cases of abnormal execution, the `evaluation details` structure's `error code` field **SHOULD** identify an error occurred during flag evaluation, with possible values `"PROVIDER_NOT_READY"`, `"FLAG_NOT_FOUND"`, `"PARSE_ERROR"`, `"TYPE_MISMATCH"`, or `"GENERAL"`.
+
+##### Requirement 1.17
+
+> If the `reason` in the `evaluation details` is set to `"ERROR"`, the sibling `error code` field **MUST** be set.
+
+##### Requirement 1.18
+
+> The `evaluation options` structure's `hooks` field denotes a collection of hooks to be executed for the respective flag evaluation.
+
+See [hooks](./hooks.md) for details.
+
+##### Requirement 1.19
+
+> No methods, functions, or operations on the client should ever throw exceptions, or otherwise abnormally terminate. Flag evaluation calls must always return the `default value` in the event of abnormal execution. Exceptions include functions or methods for the purposes for configuration or setup.
+
+##### Requirement 1.20
```

Comment:
I think we should take a harder stance here. Either the SDK supports async evaluation (and the Application Author has to plan for that) or the SDK does not support it and our spec should have explicitly defined semantics for what happens when you ask for a flag evaluation but it can't be done synchronously.

The specific scenario I'm thinking about is flagging for clientside SPAs,  where you often need to evaluate a flag at the point a page is loading, where they may well not be any locally cached flag configuration loaded yet.

Reply:
I think we need to consider languages and runtimes where asynchronous methods are not available or less common. For example, in many Java frameworks, thread-per-request is quite standard, so blocking while the flag is evaluated seems pretty reasonable there. Only a single request is blocked.

I see your concern, I just dont want to make it a hard requirement for all languages, especially ones in which blocking is common, or even idiomatic.

Reply:
Why not say that the client **should** provide both async and sync mechanisms?
One consumer will probably want to almost-always call async methods, yet another consumer will probably want to almost-always call sync methods. Forcing only async to the consumer will likely result in just a lot of unnecessary `async` (JS) or `.get()' (Java future) calls, which they will probably write a simple wrapper around - so why not make that wrapper part of the spec?

It's pretty trivial for a provider to turn an async method into a sync one by using `await` `.get()` if their vendor-specific implementation only provides async methods, or turn a sync method into an async one by wrapping with `new Promise` etc.

Reply:
> Why not say that the client should provide both async and sync mechanisms?

I don't think this is reasonable, or possible in Javascript especially, given that we have vendors SDKs with ONLY async APIs already. If a method returns a promise, there's no way to make that method synchronous and also behave predictably without blocking the thread of execution (which means blocking your entire webapp in the case of node).

You're right that it's certainly possible to make a synchronous call async by wrapping it in a promise, but you cant go the other way. I also didn't want to force languages/frameworks where blocking is reasonable NOT to block (Java). Of course, I'm not against Futures or callbacks, but its certainly not unusual for Java calls to block.

Reply:
yeah, you're right. I was thinking it is trivial to just `await` it, but of course, that makes your calling function async, then it's turtles all the way down.

so I wonder about saying "the client **should**" provide both a sync and an async method" (for those who want to call it in either form". If a vendor only provides an async method (and the language can't sync-ify it) then one option _could_ be to throw a `NotImplementedException` or similar. Though I do agree with the concept of 'our evaluation APIs shouldn't throw errors'. Hmmm.... I'll have a think. What are your thoughts in general to providing both sync and async evaluation methods?

Reply:
Ooooh, thought.
one pattern used elsewhere would be `OpenFeature.getSyncClient(...)`  and `OpenFeature.getAsyncClient(...)`. That way, if needed, the getSyncClient could throw an error (legal in the current spec) without the `client.getStringValue()` itself needing to throw the exception to go "this isn't supported". Thoughts?

Reply:
> What are your thoughts in general to providing both sync and async evaluation methods?

I think it makes perfect sense in some implementation languages, but not others. It's perfectly reasonable for Java, but not JS. Like I mentioned before, it's pretty common for JS libs not to have any sync methods at all, and for good reason.

We could change it to a **SHOULD** that would just force the implementer to consider this question, not sure how to word it, but I'll think on it and respond.

Reply:
> Ooooh, thought.
one pattern used elsewhere would be OpenFeature.getSyncClient(...) and OpenFeature.getAsyncClient(...). That way, if needed, the getSyncClient could throw an error (legal in the current spec) without the client.getStringValue() itself needing to throw the exception to go "this isn't supported". Thoughts?

I think this is a reasonable idea, and might be a good path forward. I suppose my only concern is whether it's possible that there's languages where this concept itself wouldn't be idiomatic. I don't think Golang, with it lightweight goroutine concept and channels, typically sees explicitly async APIs. In Go, it's pretty easy to start a goroutine and wait for it to write to a channel without needing to "mark" the entire stack as async, like you would for async/await in c# or JS.

Reply:
My Java implementation is blocking. 

I like the notion of a nonblocking version but I'm unsure if it should be the same repo or a different one. There are a *bunch* of methods we're already providing in the Java sdk. I'm not the most excited about doubling the count, but could do it. I'll be focusing on thread safety first.

Reply:
@toddbaert @moredip , for the languages where it doesn't make sense to have one or other of the sync or async client/methods, we could just omit those `get[A]SyncClient(...)` methods from the API.

Reply:
OK, so to make sure I understand:

In implementation languages where there's both async and sync methods, we'd recommend (require?) the API to define 2 methods:

`openfeature.getAsyncClient()` and `openfeature.getSyncClient()` (these can abnormally terminate with `NotImplemented`, since there are "configuration")

And in languages where that dichotomy doesn't really exist, it's just:

`openfeature.getClient()`

If that's what you're suggesting, I think we can reflect that in the spec, and I think it's a pretty solid plan, so it has my vote. Let's wait to see if anyone objects, and if not, I'll modify the spec accordingly.

Let me know if I've misunderstood.

Reply:
That's _very almost_ what I was suggesting 👍 .

And in languages where that dichotomy doesn't really exist, it's just:

`openfeature.getSyncClient()` or `openfeature.getAsyncClient()` (as appropriate)

IMO this keeps consistency across the different languages (and makes documentation easier for us) and is _ever so slightly_ more informative to the end-user... but TBH this slight modification is a tiny point and I'm happy with either option.

Reply:
It feels pretty heavy to have two distinct clients (sync vs async), with two distinct interfaces, **AND** different semantics about how certain error scenarios are handled.

I suspect we're "only" really talking about this because of JS support, where you can't have a blocking call. I wonder if it would be better to add a special case for that scenario - and specify that "clients which cannot support a blocking method call should instead provide an equivalent asynchronous variant" or something like that?

Reply:
> AND different semantics about how certain error scenarios are handled.

I don't think different error semantics were being proposed, though maybe I missed something. There's just a stipulation that no errors are thrown with flag evaluation, but an exception to that rule is configuration/setup code, which includes getting a client. This means, to @agentgonzo 's point, that a `NotImplemented` could be thrown.

> I suspect we're "only" really talking about this because of JS support, where you can't have a blocking call. I wonder if it would be better to add a special case for that scenario - and specify that "clients which cannot support a blocking method call should instead provide an equivalent asynchronous variant" or something like that?

Yes, that, combined with the fact with already have a provider with an async underlying SDK, is the only reason I added anything about asynchrony in the spec at all, and perhaps that was a mistake. I think for now we should merge without having this nailed down, and we can revisit this specific issue later.

Reply:
@toddbaert I'm not sure. I think we leave this as is, merge it, then discuss in an issue.

I would want to think through what a sync hook vs a sync hook might look like before I committed one way or the other, but this is good enough to make progress on imo.